scipy/special/tests/test_basic.py

```diff
@@ -3618,6 +3621,27 @@ def test_lpmn(self):
                                                       1.00000,
                                                       1.50000]])),4)
 
+    @pytest.mark.parametrize('n', [0, 1, 2, 5, 10])
+    def test_lpmn_mp(self, n):
+        @np.vectorize
+        def func(m, n, x):
+            if (abs(m) > n):
+                return 0
+
+            return float(mp.legenp(n, m, x))
+
+        rng = np.random.default_rng()
+
+        x = 0.256
+        for m in random.choices(tuple(range(n + 1)), k = 5):
```

Comment:
These `m` should be parametrized too. The point is that we don't want the whole test to fail if just one case fails, before running other cases. Same for having multiple tests like this in each iteration of the loop.

Reply:
Right, obviously I accept that we need to stick to SciPy's conventions here, but some of this stuff I actually think is generally okay. For instance, `self.subTest` is sort of for dynamic subtests.

Reply:
Yeah. It depends on the structure of the project. In what’s basically a monorepo structure like SciPy, where all of the tests run in CI even though PRs typically touch one subpackage at a time, it adds too much noise if we have tests which may fail randomly. If we had our own library just for special functions, random tests could be helpful, but I still wouldn’t have them run on every push on every PR. Maybe have them on a cron so we can find new issues over time.

Reply:
Did you notice that we have these?

https://github.com/scipy/scipy/blob/4edfcaa3ce8a387450b6efce968572def71be089/scipy/special/tests/test_mpmath.py#L1639-L1798

in `test_mpmath`. I'd forgotten about this. Also, `mpmath` tests should use the `check_mp` decorator there to avoid running if an appropriate `mpmath` version isn't installed. I guess `mpmath` tests should go in `test_mpmath.py` although we do have some other test files which use it.

Reply:
No, I didn't know that. That's very helpful, I'll have a look and see if I can move the tests there.

Reply:
My issue with these is they don't appear to be working with shapes, just scalar values.

Reply:
> My issue with these is they don't appear to be working with shapes, just scalar values.

Yeah, I think just put your tests in their own file maybe, and use the `check_mp` decorator on them.

Reply:
This whole testing thing has open a whole other kettle of fish. I've just removed the new tests, and suggest we leave this for another PR. Clearly there is general work to be done on testing in special.

Reply:
Good call. It’s part of this #20223 sdg to clean up testing anyway, so it’s on the road map, and it doesn’t really need your expertise to do it.